platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultFileCodec.kt

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl.serialize
+
+import org.gradle.internal.serialize.graph.CloseableReadContext
+import org.gradle.internal.serialize.graph.CloseableWriteContext
+import org.gradle.internal.serialize.graph.FileDecoder
+import org.gradle.internal.serialize.graph.FileEncoder
+import org.gradle.internal.serialize.graph.FilePrefixedTree
+import org.gradle.internal.serialize.graph.FilePrefixedTree.Node
+import org.gradle.internal.serialize.graph.ReadContext
+import org.gradle.internal.serialize.graph.WriteContext
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+import kotlin.concurrent.thread
+
+private const val EOF = (-1).toByte()
+private const val NODE_FINAL_START = 1.toByte()
+private const val NODE_INTERMEDIATE_START = 2.toByte()
+private const val NODE_END = 3.toByte()
+
+class DefaultFileEncoder(
+    private val globalContext: CloseableWriteContext,
+    private val prefixedTree: FilePrefixedTree
+) : FileEncoder {
+
+    override fun writeFile(writeContext: WriteContext, file: File) {
+        writeContext.writeSmallInt(prefixedTree.insert(file))
+    }
+
+    override fun close() {
+        globalContext.use {
+            it.writePrefixedTreeNode(prefixedTree.compress(), null)
+            it.writeByte(EOF)
+        }
+    }
+
+    private fun WriteContext.writePrefixedTreeNode(node: Node, parent: Node?) {
```

Comment:
All this thing is about as less disk as possible for metadata, so we encode parent/child relationship with byte markers with further using stack for the reading. Something like Balanced brackets challenge.

Also we encode node index using delta encoding - we store a delta of `node.index-parent.index` assuming that their indexes are close, and the result of deduction is small enough to fit byte size